Collections/CountedHashSet.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System.Collections.Generic;
+
+namespace Anvil.CSharp.Collections
+{
+    /// <summary>
+    /// A hash set that maintains a balance of add/remove calls and only removes the item when <see cref="Remove"/> has
+    /// been called as many times as <see cref="Add"/>.
+    /// </summary>
+    /// <typeparam name="T">The type of items to store.</typeparam>
+    public class CountedHashSet<T>
+    {
+        private readonly Dictionary<T, int> m_Items;
+
+        /// <summary>
+        /// A collection of items that have been added to the set.
+        /// </summary>
+        public Dictionary<T, int>.KeyCollection Items
+        {
+            get => m_Items.Keys;
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="CountedHashSet{T}"/>.
+        /// </summary>
+        public CountedHashSet()
+        {
+            m_Items = new Dictionary<T, int>();
+        }
+
+        /// <summary>
+        /// Add an item to the set or increase the count if it already exists.
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <returns>True if the item is new to the set.</returns>
+        public bool Add(T item)
+        {
+            if (!m_Items.ContainsKey(item))
+            {
+                m_Items.Add(item, 1);
+                return true;
+            }
+
+            m_Items[item]++;
+            return false;
```

Comment:
Done

Reply:
While setting up a new Anvil project and trying different Unity versions, I've discovered that [`Dictionary.TryAdd()`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.tryadd?view=net-6.0) is a .NET Standard 2.1 feature, meaning Anvil [requires Unity 2021.2+](https://youtu.be/eILAjaBoV2Y?t=285).

Was that a conscious decision? If so, I'll update the README to reflect the new minimum .NET and Unity versions.

Reply:
oh that's a good catch and not a conscious decision. @jkeon thoughts on maintaining .NET 4.x compatibility?

Reply:
I think that as a team of the three of us, maintaining backwards compatibility is a bit of chore. 

If any of us had active projects that required older versions, then yes we should do so and that's fine but otherwise I think it's fine to push forward to the latest and greatest as we can. 

I think our biggest holdup right now is Unity's ECS 0.51 so we're stuck on Unity 2021.3.4+ at minimum which gives us C# 9.0 (https://docs.unity3d.com/Manual/CSharpCompiler.html) 

Other projects in Unity or .Net Console are all higher versions. 


If for some reason any of us has an older project that we want to keep on an older version, we could always add some #ifdefs but I don't think we do?

Reply:
I have no plans or reason to use older Unity/C#/.NET versions, so I'm on board - it took barely any effort to get Anvil working in Unity 2019, should anyone require that at some point.

I tried updating the README files for both csharp and unity repos, but wasn't sure where (if anywhere) to mention the minimum Unity and .NET versions required by each. Dependencies? Project setup?

Reply:
I'd modify the existing dependency bullet points in both READMEs. 
I'd avoid any mention of Unity in the `anvil-csharp-core` README and, if possible, just change the existing .NET mention.

For `anvil-unity-core` you could probably just add a version number to the existing Unity dependency and if it needs to be configured a certain way to support `anvil-csharp-core` then  mention it there.
Ex: 
 - [anvil-csharp-core](https://github.com/decline-cookies/anvil-csharp-core) - Requires Player Settings set to .Net Standard 2.1